### ISSUE TITLE
Duplicate pull requests for .Net projects chained via project reference

### ISSUE THREAD

**Issue by krlxgrn** [dependabot/dependabot-core#1408]
When using the dependabot config file with multiple projects in a repository, I'm seeing multiple pull requests for the same package update.

To reproduce create a new .NET solution with two projects, A and B where B depends on A via a project reference.

Add an outdated package to project A.

Setup the dependabot config file so that each project is listed separately and 2 pull requests will be opened.

One will say `Bump Package X in /A` and the second will say `Bump Package X in /B` even though they are both updating the same package in the same project.

I would expect this to just open one pull request and additionally it's confusing seeing the PR title saying `Bump Package X in /B` when the code change is not changing project B directly at all.

I made an example repo that shows the issue, hopefully that helps https://github.com/k-arl/DependabotDuplication

**Comment by kilbergr** [comment]
Related but not identical.
In my case I see duplicate PRs for the same change when I configure dependabot to watch multiple directories for the same language (Python).
Exact same behavior as above.

You can see example here:
https://github.com/trussworks/circleci-docker-primary/pull/46
and 
https://github.com/trussworks/circleci-docker-primary/pull/47
are for the same change.

This is the config for the repo: https://github.com/trussworks/circleci-docker-primary/blob/master/.dependabot/config.yml

**Comment by feelepxyz** [comment]
@kilbergr you should be able to remove this entry in the your config: https://github.com/trussworks/circleci-docker-primary/blob/master/.dependabot/config.yml#L17-L23 as the first one covers both requirement.txt files (Dependabot searches for txt files one level deep).

**Comment by kilbergr** [comment]
Oh great! I was looking for documentation about whether that's a thing--i.e. how deep in the tree dependabot goes, but I couldn't find. Does documentation about that exist? Thanks!

**Comment by feelepxyz** [comment]
@kilbergr we're missing a lot of documentation on all of these gotchas, something we're slowly writing up and plan to publish. The best bet is currently looking at the code or asking here 🤠 https://github.com/dependabot/dependabot-core

**Comment by mwaddell** [comment]
Same as #3856 I think

**Comment by Jericho** [comment]
I just setup dependabot for the first time and it created several duplicate PRs, exactly as described by @krlxgrn. In my case it's even worse because my solution contains 4 projects, 3 of which reference the first one. Dependabot correctly detected 5 outdated NuGet package references in the first project and created the expected 5 PRs, but it also created 5 duplicate PRs for the second project, 5 duplicate PRs for the third project and 5 duplicate PRs for the fourth project. 

In other words, instead of the 5 expected PRs, I ended up with 20 PRs, 15 of which are duplicates.

**Comment by jeffwidman** [comment]
This only happens in a monorepo, right?

I'm not super familiar with `.Net` so unclear if the "projects" terminology refers to disparate apps/libraries, or to submodules for a single app. But it sounds like you've got a couple of libraries sitting alongside an actual app that pulls in those libraries...

**Comment by jeffwidman** [comment]
Is this a duplicate??
* https://github.com/dependabot/dependabot-core/issues/1595

**Comment by Jericho** [comment]
> This only happens in a monorepo, right?

Yes, it's a single GitHub repository.

> I'm not super familiar with .Net so unclear if the "projects" terminology refers to disparate apps/libraries, or to submodules for a single app. But it sounds like you've got a couple of libraries sitting alongside an actual app that pulls in those libraries...

In .Net you have a solution (a file with `.sln` extension) which contains links to one or multiple projects (if you use C#, your project files will have `.csproj` extension, if you use VB.NET the extension is `.vbproj` and if you use F# the extension is `.fsproj`). These project can reference nuget packages (these are the references we expect dependabot to keep up-to-date) but they can also reference each other. For example, your solution could include a library project and a unit tests project. In this example, the tests project would reference the library. There are a multitude of other project types (console app, web app, web API, winform, blazor, etc.), the two that I mentioned are some of the ones I personally use most often.

In fact, feel free to look at the [GitHub repo where I experienced this problem](https://github.com/Jericho/StrongGrid), you'll see exactly what I'm talking about. While you're there, look at all the [PRs that debendabot created](https://github.com/Jericho/StrongGrid/pulls?q=is%3Apr+is%3Aclosed+author%3Aapp%2Fdependabot) on March 19 2022, you'll see the numerous duplicates.

When you look at the dupes, please pay attention to one detail in particular that may be confusing (it confused me at first before I realized what was going on):
- Dependabot noticed an outdated nuget package reference in my library project and correctly raised this PR: `Bump System.Text.Encoding.CodePages from 5.0.0 to 6.0.0 in /Source/StrongGrid`
- Dependabot looked at my unit testing project (which references my library project) and noticed that my library has an outdated nuget package reference and raised a duplicate PR: `Bump System.Text.Encoding.CodePages from 5.0.0 to 6.0.0 in /Source/StrongGrid.UnitTests`. The confusing part is the title of this duplicate PR makes it sound like the outdated package reference is in the unit testing project.
- Dependabot looked at my third project which I use for benchmarking purposes (this project also references my library) and raised another duplicate PR: `Bump System.Text.Encoding.CodePages from 5.0.0 to 6.0.0 in /Source/StrongGrid.Benchmark`. Again, the title of the PR implies that the nuget reference is in the bechmarking project when in fact it's not.

> Is this a duplicate??

It does indeed sound very similar.

**Comment by Jericho** [comment]
> > Is this a duplicate??
> 
> It does indeed sound very similar.

Actually, I take that back. I re-read the issue you linked and I think they are describing a completely different situation. What they are talking about is the case where project A and project B both reference nuget package Z. In this situation dependabot raises two PRs to upgrade the package Z references which they find "noisy". In their scenario the multiple PRs are valid, it's just that it would be more convenient if the various commits were grouped in a single PR. If you look at the screenshot in the July 2020 comment, you'll see 8 PRs to update a package to the latest version. My understanding is that there are indeed 8 references to the same package that need to be upgraded, these are not duplicates.

In my scenario, there are multiple duplicate PRs. So, it's not the same issue at all.

**Comment by abdulapopoola** [comment]
**Update**: We've started doing some grouped updates work! This particular issue might not be part of the first ship but if you want to track our updates, do follow #1190.

**Comment by Jericho** [comment]
Grouping several commits into a single PR is a nice improvement and will resolve #1595 but it will not help with the problem presented in this issue. @krlxgrn originally brought up the issue and @jeffwidman and I had a subsequent discussion where I attempted to clearly present the problem leading dependabot to create duplicates.

At the risk of repeating what's already been said many times in this discussion: the problem is that dependabot creates multiple PRs for the same update. The solution is not to group the multiple duplicates in a single PR. The solution is to avoid creating duplicates in the first place. I urge you to review the [PRs created by dependabot in my repo on March 19 2022](https://github.com/Jericho/StrongGrid/pulls?q=is%3Apr+is%3Aclosed+author%3Aapp%2Fdependabot) in order to witness the several duplicates created by dependabot.

**Comment by jeffwidman** [comment]
Thanks @Jericho and we do understand what you're asking for. 

While this issue clearly affects you and a couple of other folks, the overall impact across our userbase appears to be pretty low (for example, currently no 👍 on the original comment), so it's unlikely to be prioritized by the core team anytime soon. 

However, the great thing is this is all open source, so we're happy to review a PR and answer any questions along the way. But I just want to be transparent that this is unlikely to move forward until someone from the community steps up to fix it.

**Comment by abdulapopoola** [comment]
Closing out as this has not seen any activity in recent times; please feel free to reopen if otherwise.
